hack/apidiff.sh

```diff
@@ -199,4 +228,93 @@ for d in "${targets[@]}"; do
     compare "${d}" "${KUBE_TEMP}/before/$(output_name "${d}")" "${KUBE_TEMP}/after/$(output_name "${d}")"
 done
 
+# tryBuild checks whether some other project builds with the staging repos
+# of the current Kubernetes directory.
+#
+# shellcheck disable=SC2317 # "Command appears to be unreachable" - gets called indirectly.
+tryBuild () {
+    local build="$1"
+
+    # Replace all staging repos, whether the project uses them or not (playing it safe...).
+    local repo
+    for repo in $(cd staging/src; find k8s.io -name go.mod); do
+        local path
+        repo=$(dirname "${repo}")
+        path="$(pwd)/staging/src/${repo}"
+        (
+            cd "$build"
+            go mod edit -replace "${repo}"="${path}"
+        )
+    done
+
+    # We only care about building. Breaking compilation of unit tests is also
+    # annoying, but does not affect downstream consumers.
+    (
+        cd "$build"
+        rm -rf vendor
+        go mod tidy
+        go build ./...
+    )
+}
+
+if [ $res -ne 0 ]; then
+    cat <<EOF
+
+Some notes about API differences:
+
+Changes in internal packages are usually okay.
+However, remember that custom schedulers
+and scheduler plugins depend on pkg/scheduler/framework.
+
+API changes in staging repos are more critical.
+Try to avoid them as much as possible.
+But sometimes changing an API is the lesser evil
+and/or the impact on downstream consumers is low.
+Use common sense and code searches.
+EOF
+
+    if [ ${#builds[@]} -gt 0 ]; then
+
+cat <<EOF
+
+To help with assessing the real-world impact of an
+API change, $0 will now try to builds code in
+${builds[@]}.
+EOF
+
+        if [[ "${builds[*]}" =~ controller-runtime ]]; then
```

Comment:
How is the controller-runtime version/branch selected that is used? (and which one would be used?)

Reply:
This needs to be selected in the Prow job. I would use "master" there because that is most likely to work with Kubernetes "master", which the job is for.

There is of course no guarantee that this builds even without any API changes in a PR. That's what the `This may or may not be because of the API changes!` warning is about.

But that's okay. The Prow job only provides additional information which needs to be interpreted. It cannot be merge-blocking.

Reply:
Sounds good. So we can use different CR branches in different k/k branches if necessary.

I think it makes sense as a goal to have CR main compile with k/k master. Will require CR to adopt to k/k changes here and there, but that's what we did in the past as well when we bumped to k/k alpha / beta / ... releases

Reply:
> So we can use different CR branches in different k/k branches if necessary.

We currently don't fork `pull-kubernetes-apidiff` for old releases because backports really shouldn't contain anything which changes an API and reviewers are aware, so we don't need a job for it.

Reply:
I don't think this should live in Kubernetes, since this can be aded on the external projects.

Per example, we had the same problem in cloud provider gcp and I added a periodic job to build against master and a presubmit in the cloud provider gcp to not break the compatibility 

https://github.com/search?q=repo%3Akubernetes%2Ftest-infra+%22add-kubernetes-to-workspace.sh%22&type=code

https://github.com/kubernetes/cloud-provider-gcp/blob/c2d05af3b14f885e2560139dd6fb4581ceb01733/e2e/add-kubernetes-to-workspace.sh#L20-L41

I don't think we should create this dependency on controller-runtime in this repo

Reply:
The purpose of `pull-kubernetes-apidiff` is to inform Kubernetes developers about the impact of making an API change *before* committing that change to the repo. We've had lots of discussions in different PRs about "this is an API break,what's the real-world impact?". Sometimes it's okay to change the API, sometimes it's not.

controller-runtime was frequently sited as an example that should be looked at because it serves as a stand-in for controllers built on top of our APIs. This PR simplifies making that check.

> I don't think we should create this dependency on controller-runtime in this repo, 

It's a very "soft" dependency: we get some information, but nothing depends on that information.

> since it is perfectly possible for controller runtime to have this coverage today without adding anything here

I don't want this for the controller-runtime, I want this for Kubernetes :smile: 

/cc @sttts

Reply:
Let me also point out a difference between "cloud provider gcp" and controller-runtime: the former is code from a vendor, the latter is a Kubernetes SIG project. I wouldn't want to favor one vendor over another by testing against their code, but testing some other code maintained under the Kubernetes umbrella seems fair to me.

Reply:
maybe I misunderstood, what I try to mean is that we can not have a job that depends on controller-runtime and create a bidirectional dependency, because as one direction is clear we want to be sure do not break anyone, the other direction can happen, and we don't want to have to modify kubernetes code if controller runtime breaks dependency against us

Reply:
> Let me also point out a difference between "cloud provider gcp" and controller-runtime: the former is code from a vendor, the latter is a Kubernetes SIG project.

Er, they're both from SIGs, actually. Cloud Provider GCP is owned by SIG Cloud Provider.

>  I wouldn't want to favor one vendor over another by testing against their code, but testing some other code maintained under the Kubernetes umbrella seems fair to me.

Every PR runs cloud-provider-gcp and cloud-provider-aws because that's where we have sufficient resources. This is unavoidable and we need to know that they work so we can test other things.

I don't think we should use either. We should be determining in this repo what APIs are stable or not, we should not skip including methods just because we didn't select a project that uses it.

Reply:
> Er, they're both from SIGs, actually. Cloud Provider GCP is owned by SIG Cloud Provider.

Thanks for the correction.

> what I try to mean is that we can not have a job that depends on controller-runtime and create a bidirectional dependency

Don't worry, that's not what this PR does. If controller-runtime doesn't compile with current Kubernetes master, then that's okay. It does not force us to change Kubernetes. That's because running `pull-kubernetes-apidiff` is purely for informational purposes (not run by default, not merge-blocking). How to interpret the results is up to the developer who invoked it.

We can also pick other projects to build with. The script supports that. Coverage is never going to be "complete", but some answer with controller-runtime is better than none and perhaps there are other projects that give us a similar "bang for the buck".

Reply:
ACK.

I still think we should be deciding "these APIs are intended to be stable" and tagging them (maybe a list of packages?), and then diffing those APIs.

Reply:
Every exported API in a staging repo is something that someone could depend upon. I don't think it would make of a difference now if we were to declare some of them as "stable" and "not stable".

Reply:
I think the questions is if ones are going to be "more stable" than others, if all the same there is no much point in checking against other projects, or I miss the benefit

Reply:
I think we have to keep using common sense here instead of defining some APIs as stable and others as unstable. The bar for changing Go APIs in staging repos should always be high because of the risk that we break some unknown downstream consumer.

In cases where maintaining API compatibility seems to troublesome (extra work, ugly API, etc.), then we should gather information about the impact (consult experts, code searches, or, as in this PR, trial builds) before making a decision.

I've seen PRs where APIs were broken without such due diligence or where changes were proposed that were said to restore API compatibility without validating that they really do (they didn't). My hope is that by making `pull-kubernetes-apidiff` more informative, we will do better in the future.

Reply:
+100